### PR TITLE
Add interface and schema detection for contracts

### PR DESCRIPTION
**Pull request by fhildeb**
This PR introduces interface ID and schema key checkups as standalone modular function within the library that can be called directly on the _ERC725 object_ or an _custom smart contract address as parameter_. For convenience, the detection tool can not only handle interface IDs and schema Keys, but also a variety of names. 

### Functions:
```
ERC725.supportsInterface(interfaceId or interfaceName, options) > bool
myErc725.supportsInterface(interfaceId or interfaceName) > bool

ERC725.supportsSchema(schemaKey or schemaName, options) > bool
myErc725.supportsSchema(interfaceId or schemaName) > bool
```

### Default supported interface names: 
```
ERC1271,
ERC725X,
ERC725Y,
LSP0ERC725Account,
LSP1UniversalReceiver,
LSP1UniversalReceiverDelegate
LSP6KeyManager,
LSP7DigitalAsset,
LSP8IdentifiableDigitalAsset,
LSP9Vault
```

### Default supported schema names: 
```
LSP3UniversalProfile,
LSP4DigitalAssetMetadata,
LSP9Vault,
SupportedStandard:LSP3UniversalProfile,
SupportedStandard:LSP4DigitalAssetMetadata,
SupportedStandard:LSP9Vault
```

When using one of the default schema names, the corresponding schema is also loaded directly from the library.

### References

#205 
#235 

### Pending

- [ ] Write documentation
- [ ] Add value/mockup tests

### CODE REVIEW

**Comment by CallumGrindle** [comment]
Please remove the package-lock.json from this PR. There are no changes to package.json so we don't expect any changes to package-lock

File: src/lib/constants.ts [hunk 1]
```diff
@@ -169,3 +169,64 @@ export const LSP6_DEFAULT_PERMISSIONS = {
...
+          type: 'bool',
+        },
+      ],
+      stateMutability: 'view',
+      type: 'function',
+    },
+  ],
+};
+
+export const IPFS_GATEWAY_BASE_URL = 'https://ipfs.io/ipfs/'; // 'https://2eff.lukso.dev/ipfs/' not working;
```
Comment (CallumGrindle) [inline src/lib/constants.ts:197]: Is there not already a default ipfs gateway value somewhere? Why not use that?

File: src/lib/constants.ts [hunk 2]
```diff
@@ -169,3 +169,64 @@ export const LSP6_DEFAULT_PERMISSIONS = {
...
+
+export enum LSPType {
+  LSP3UniversalProfile = 'LSP3',
```
Comment (CallumGrindle) [inline src/lib/constants.ts:200]: We should use full LSP names

File: src/lib/constants.ts [hunk 3]
```diff
@@ -169,3 +169,64 @@ export const LSP6_DEFAULT_PERMISSIONS = {
...
+  LSP1UniversalReceiver: '0x6bb56a14',
+  LSP1UniversalReceiverDelegate: '0xa245bbda',
+  LSP6KeyManager: '0xc403d48f',
+  LSP7DigitalAsset: '0x5fcaac27',
+  LSP8IdentifiableDigitalAsset: '0x49399145',
+  LSP9Vault: '0xfd4d5c50',
+  ClaimOwnership: '0xa375e9c6',
+};
+
+export interface LspTypeOption {
```
Comment (CallumGrindle) [inline src/lib/constants.ts:229]: ideally interfaces would go in some `interfaces` file as they are not really constants

File: src/lib/constants.ts [hunk 4]
```diff
@@ -169,3 +169,64 @@ export const LSP6_DEFAULT_PERMISSIONS = {
...
+  ERC721: '0x80ac58cd',
+  ERC721Metadata: '0x5b5e139f',
+  ERC777: '0xe58e113c',
+  ERC1155: '0xd9b67a26',
+  ERC725X: '0x44c028fe',
+  ERC725Y: '0x714df77c',
+  LSP0ERC725Account: '0xeb6be62e',
+  LSP1UniversalReceiver: '0x6bb56a14',
+  LSP1UniversalReceiverDelegate: '0xa245bbda',
+  LSP6KeyManager: '0xc403d48f',
```
Comment (CallumGrindle) [inline src/lib/constants.ts:222]: Handling versioning for these worries me a bit, what about having an object for these containing the versions or smth?

File: src/index.ts [hunk 5]
```diff
@@ -673,6 +682,98 @@ export class ERC725 {
   encodeKeyName(keyName: string, dynamicKeyParts?: DynamicKeyParts): string {
     return encodeKeyName(keyName, dynamicKeyParts);
   }
-}
 
+  baseConfig = { ipfsGateway: IPFS_GATEWAY_BASE_URL };
```
Comment (CallumGrindle) [inline src/index.ts:686]: Can you not use the ipfs config of the ERC725 class rather than recreate it here?

File: src/index.ts [hunk 6]
```diff
@@ -673,6 +682,98 @@ export class ERC725 {
...
+
+  lspTypeOptions: Record<Exclude<LSPType, LSPType.Unknown>, LspTypeOption> = {
```
Comment (CallumGrindle) [inline src/index.ts:688]: What is this object? It just contains the interface id which we already have in a constant and the SupportedStandard ABI which should be the same for each standard anyway? I think we can simplify this and do without this object

File: src/index.ts [hunk 7]
```diff
@@ -673,6 +682,98 @@ export class ERC725 {
...
+      return results[0];
+    } catch (error) {
+      return null;
+    }
+  }
+
+  async detectLSPInterface(
+    contractAddress: string,
+    lspType: LSPType,
+    web3Provider: Web3,
```
Comment (CallumGrindle) [inline src/index.ts:734]: This means we cant use an ethereum provider. You can use the provider that is set on the ERC725 class

File: src/index.ts [hunk 8]
```diff
@@ -673,6 +682,98 @@ export class ERC725 {
...
+    }
+
+    // EIP-165 detection
+    const contract = new web3Provider.eth.Contract(
+      COMMON_ABIS.supportsInterface as AbiItem[],
+      contractAddress,
+    );
+
+    // Check if the contract implements the LSP interface ID
+    let doesSupportInterface: boolean;
```
Comment (CallumGrindle) [inline src/index.ts:747]: what about `isInterfaceSupported`?

File: src/index.ts [hunk 9]
```diff
@@ -673,6 +682,98 @@ export class ERC725 {
...
+    try {
+      doesSupportInterface = await contract.methods.supportsInterface(
+        this.lspTypeOptions[lspType],
+      ).call;
```
Comment (CallumGrindle) [inline src/index.ts:751]: Does this not need to be called? It looks like this sets the `doesSupportInterface` to the `call` function. Might be a web3 thing. Can you check?

**Review by CallumGrindle — CHANGES_REQUESTED** [review]
Nice function. Though I don't think we should be assuming these standards necessarily coexist and e.g. that an LSP0ERC725Account necessarily supports LSP3. This will lead to confusing outcomes as I could have a valid anon ERC725Account and this function would tell me it is not supported as it does not implement LSP3. 

We should treat each standard separately as they are modular and each standard can be used independently and in different combinations.

e.g. an LSP7 contract does not need to implement LSP4 and LSP4 can be used on a contract which is not LSP7 or LSP8. 

I think decoupling these standards will simplify this function for you as you will not need to check these metadata standards unless LSP3,LSP4,LSP9 is specifically being asked for

File: src/lib/constants.ts (hunk 1 above)
Comment (fhildeb) [reply src/lib/constants.ts:197]: for some reason, the cloudflare IPFS does not work sometimes for assets on l16. had this issue with Jean several times, so i just thought I go with how we handle it in the wallet.

File: src/index.ts (hunk 6 above)
Comment (fhildeb) [reply src/index.ts]: We wanted to have one object that can hold interfaceId but also Schema, so the function works pretty generic, and we just check if its a schema or interface standard- and then we basically map everything thats needed onto the LSPs.

File: src/index.ts (hunk 8 above)
Comment (fhildeb) [reply src/index.ts]: we can leave it out, initially i planned on having a boolean but in the latest version due to modularity, we just return plain values

File: src/index.ts (hunk 9 above)
Comment (fhildeb) [reply src/index.ts]: it will just return the value of the supportsInterface function

**Comment by fhildeb** [comment]
Split into multiple PRs:

#243 --> merged
#244 --> open discussion

Closed.

**Pull request closed without merging**